Add x-amz-storage-class support

nitisht opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
This adds configurable data and parity options on a per object
basis. To use variable parity

- Users can set environment variables to cofigure variable
parity

- Then add header x-amz-storage-class to putobject requests
with relevant storage class values

## Motivation and Context
Fixes #4997

## How Has This Been Tested?
Manually and mint runs

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

krisis commented on cmd/common-main.go:
```diff
@@ -106,4 +106,38 @@ func handleCommonEnvVars() {
 	}
+
+	// Validate and store the storage class env variables only for XL/Dist XL setups
+	if globalIsXL || globalIsDistXL {
```
`globalIsXL` is true if `globalIsDistXL` is true. So, it is sufficient to check if `globalIsXL` is true to support storage classes.

krisis commented on cmd/config-migrate.go:
```diff
@@ -1704,3 +1708,110 @@ func migrateV20ToV21() error {
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("Unable to load config version ‘20’. %v", err)
```
s/20/21

fwessels replied:
'20' --> '21'

fwessels commented on cmd/common-main.go:
```diff
@@ -106,4 +106,38 @@ func handleCommonEnvVars() {
+			fatalIf(err, "Invalid value set in environment variable %s.", reducedRedundancyStorageClass)
+		}
+
+		// Validation is done after parsing both the storage classes. This needed because we need one
```
This needed --> This is needed

fwessels commented on cmd/config-current.go:
```diff
@@ -103,6 +104,47 @@ func (s *serverConfig) SetBrowser(b bool) {
+	rrsStorageClass, err = mustParseStorageClass(s.StorageClass.RRS)
+	fatalIf(err, "Invalid value %s set in config.json", s.StorageClass.RRS)
+
+	// Validation is done after parsing both the storage classes. This needed because we need one
```
Same typo (maybe extract out common code??)

fwessels commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,255 @@
+		return storageClass{}, err
+	}
+
+	scPtr := &storageClass{
```
Wy take pointer and return *scPtr in line 83 -- might just as well use 'in value' (or better yet, store immediately in sc return parameter)

nitisht replied:
agreed.

fwessels commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,255 @@
+		parity = totalDisks / 2
+		data = totalDisks / 2
+	}
+	return data, parity
```
you can compute `data` here once because for all cases it must necessarily be `totalDisks - parity` (this would also mimize risk of intoducing errors when new class would be introduced with somebody making an error in computing data for a particular case

nitisht replied:
good idea! done.

fwessels commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,255 @@
+	return blocks, max
+}
+
+// commonDataBlocks returns the maximally occurring datablocks and its count from xlMetaV1 slice.
```
datablocks --> dataBlocks

fwessels commented on cmd/xl-v1-healing.go:
```diff
@@ -535,6 +535,16 @@ func healObject(storageDisks []StorageAPI, bucket, object string, quorum int) (i
+
+	// get Quorum for this object
+	readQuorum, _, err := objectQuorumFromMeta(xl, partsMetadata, errs)
+
```
remove empty line

fwessels commented on cmd/xl-v1-multipart.go:
```diff
@@ -641,7 +649,15 @@ func (xl xlObjects) PutObjectPart(bucket, object, uploadID string, partID int, d
+
+	// get Quorum for this object
+	_, writeQuorum, err := objectQuorumFromMeta(xl, partsMetadata, errs)
+
```
Empty new line

fwessels commented on cmd/xl-v1-multipart.go:
```diff
@@ -904,7 +920,15 @@ func (xl xlObjects) CompleteMultipartUpload(bucket string, object string, upload
+
+	// get Quorum for this object
+	_, writeQuorum, err := objectQuorumFromMeta(xl, partsMetadata, errs)
+
```
Remove empty line

fwessels commented on cmd/xl-v1-multipart.go:
```diff
@@ -1108,16 +1129,29 @@ func (xl xlObjects) cleanupUploadedParts(bucket, object, uploadID string) error
+
+	// get Quorum for this object
+	_, writeQuorum, err := objectQuorumFromMeta(xl, partsMetadata, errs)
+
```
Empty line

fwessels commented on cmd/xl-v1.go:
```diff
@@ -145,9 +142,9 @@ func newXLObjects(storageDisks []StorageAPI) (ObjectLayer, error) {
-	// READ and WRITE quorum is always set to (N/2) number of disks.
-	xl.readQuorum = readQuorum
-	xl.writeQuorum = writeQuorum
+	// READ and WRITE quorum is always set to (N/2) number of disks for metadata files.
```
write quorum is actually N/2+1

fwessels commented on cmd/xl-v1-object.go:
```diff
@@ -399,7 +415,7 @@ func rename(disks []StorageAPI, srcBucket, srcEntry, dstBucket, dstEntry string,
 	wg.Wait()
 
-	// We can safely allow RenameFile errors up to len(xl.storageDisks) - xl.writeQuorum
+	// We can safely allow RenameFile errors up to len(xl.storageDisks) - writeQuorum
```
Maybe make naming op quorum parameters in line 387 explicit, thus `writeQuorum` (so as to avoid any potential confusion).

nitisht replied:
done

krisis commented on cmd/common-main.go:
```diff
@@ -113,4 +113,38 @@ func handleCommonEnvVars() {
+
+		// Check for environment variables and parse into storageClass struct
+		if ssc := os.Getenv(standardStorageClass); ssc != "" {
+			globalStandardStorageClass, err = mustParseStorageClass(ssc)
```
must* functions in our codebase expect the parsing (or other actions) to succeed. If the action fails the functions calls os.Exit(1) or panic. So, mustParseStorageClass shouldn't return error on failure, in fact its signature should be `func mustParseStorageClass(s string) storageClass`

krisis commented on cmd/globals.go:
```diff
@@ -158,6 +157,13 @@ var (
 	// Keep connection active for clients actively using ListenBucketNotification.
 	globalSNSConnAlive = 5 * time.Second // Send a whitespace every 5 seconds.
+
+	// Storage classes
```
a single line comment for the following 3 variables is needed.

krisis commented on cmd/handler-utils.go:
```diff
@@ -127,6 +131,15 @@ func extractMetadataFromHeader(header http.Header) (map[string]string, error) {
 			metadata[supportedHeader] = header.Get(supportedHeader)
 		}
 	}
+	// Save StorageClass header
```
can't we append supportedStorageClass header(s) to supportedHeaders list and unify the saving of headers to a single range-loop?

nitisht replied:
appending `supportedStorageClass` to `supportedHeaders` in `extractMetadataFromHeader` causes a data race in unit tests. We'll probably need to add a separate method to append all the supported headers and that can be called in `extractMetadataFromHeader`.

nitisht replied:
added this method @krisis PTAL

krisis commented on cmd/handler-utils.go:
```diff
@@ -63,6 +63,10 @@ var supportedHeaders = []string{
 }
 
+var supportedStorageClassHeaders = []string{
+	"x-amz-storage-class",
```
nit: should use amzStorageClass string constant defined elsewhere.

krisis commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,252 @@
+		return errors.New("Reduced redundancy storage class not supported for " + strconv.Itoa(len(globalEndpoints)) + " disk setup")
+	}
+
+	// RRS parity disks should be greater than or equal to 2. Parity below 2 is not recommended.
```
nit: comment should refer to minimumParityDisks instead of 2.

krisis commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,252 @@
+		parity = totalDisks / 2
+	}
+	// data is always totalDisks - parity
+	data = totalDisks - parity
```
this can be simplified to `return totalDisks - parity, parity`

krisis commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,252 @@
+
+	// get the latest update XL Metadata
+	latestXLMeta := listLatestXLMeta(partsMetaData, errs)
+
```
this can be simplified as,

```go
emptyMeta := xlMetaV1{}
count := 0
for _, xlMeta := range latestXLMeta {
  if xlMeta == emptyMeta {
    continue
  }
  count ++
  if validXLMeta == emptyMeta  {
    validXLMeta = xlMeta
  }
}
if validXLMeta == emptyMeta || count < validXLMeta.Erasure.DataBlocks {
  return 0, 0, errXLReadQuorum
}
return validXLMeta.Erasure.DataBlocks, validXLMeta.Erasure.DataBlocks+1, nil
```

krisis commented on cmd/storage-class_test.go:
```diff
@@ -0,0 +1,280 @@
+			Scheme: "EC",
+			Parity: 4},
+			errors.New("Unsupported scheme AB. Supported scheme is EC")},
+		{4, "AB:4:5", storageClass{
```
it is less confusing for the reader if we have exactly one thing wrong about the input being parsed. E.g "AB:4:5" has both scheme and number of sections wrong. Since scheme related failure test case is already present, we could use "EC:4:5" as the invalid input to this unit test.

harshavardhana commented on cmd/common-main.go:
```diff
@@ -113,4 +113,38 @@ func handleCommonEnvVars() {
+		// storage class value to deduce the correct value of the other storage class.
+		if globalRRStorageClass.Scheme != "" {
+			err := validateRRSParity(globalRRStorageClass.Parity, globalStandardStorageClass.Parity)
+			if err != nil {
```
```
if err != nil
``` 

is not needed for `fatalif()`

harshavardhana commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,213 @@
+// Validates the parity disks for Reduced Redundancy storage class
+func validateRRSParity(rrsParity, ssParity int) (err error) {
+
+	// Reduced redundancy storage class is not supported for 4 disks XL setup.
```
Rename all references to XL to erasure coded..

harshavardhana commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,213 @@
+
+	// Reduced redundancy storage class is not supported for 4 disks XL setup.
+	if len(globalEndpoints) == 4 && rrsParity != 0 {
+		return errors.New("Reduced redundancy storage class not supported for " + strconv.Itoa(len(globalEndpoints)) + " disk setup")
```
You can use `fmt.Errorf`

harshavardhana commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,213 @@
+
+	// RRS parity disks should be greater than or equal to minimumParityDisks. Parity below minimumParityDisks is not recommended.
+	if rrsParity < minimumParityDisks {
+		return errors.New("Reduced redundancy storage class parity should be greater than or equal to " + strconv.Itoa(minimumParityDisks))
```
You can use `fmt.Errorf`

harshavardhana commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,213 @@
+	// Now loop through latestXLMeta and to get a count of valid xlMeta(s)
+	for _, xlMeta := range latestXLMeta {
+		if reflect.DeepEqual(xlMeta, emptyXLMetaV1) {
+			continue
```
If you use continue you don't need the else block. 

```
if reflect.DeepEqual(xlMeta, emptyXLMetaV1) {
     continue
}
count++
validXLMeta = xlMeta
```

fwessels replied:
Or negate condition of if statement:
```
if !reflect.DeepEqual(xlMeta, emptyXLMetaV1) {
    count++
    validXLMeta = xlMeta
}
```

fwessels replied:
On a second thought, if we let `listLastXLMeta` return a count then we can remove this loop and we also do not need the (expensive) `reflect.DeepEqual` call.

listLastXLMeta would end like this:
```
	count := 0
	for index, t := range modTimes {
 		if t == modTime {
 			latestXLMeta[index] = partsMetadata[index]
 			count++
 		}
 	}
 	// This is a slice with either partsMetaData that were updated the latest
 	// or empty partsMetaData
 	return latestXLMeta, count 
 }
```

nitisht replied:
>  if we let listLastXLMeta return a count then we can remove this loop and we also do not need the (expensive) reflect.DeepEqual call.

Good idea, will do this.

harshavardhana commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,213 @@
+
+	// Simple majority of dataBlocks/parityBlocks from xl.json is not enough. We need to establish that
+	// at least dataBlocks number of xlMeta is available. This means count should be greater than or
+	// equal to dataBlocks field of validXLMeta. If not we through read quorum error.
```
s/If not we through/If not we throw/

harshavardhana commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,213 @@
+}
+
+// returns readQuorum and writeQuorum based on number of blocks
+func objectQuorumFromDrives(dataBlocks int) (readQuorum, writeQuorum int) {
```
Not sure if we need this function?

fwessels replied:
Yes, let's collapse into function above.

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -264,7 +261,9 @@ func getStorageInfo(disks []StorageAPI) StorageInfo {
-	storageInfo.Backend.WriteQuorum = xl.writeQuorum
+	// return the default quorum data
+	// per object disk info is available in object meta data
+	storageInfo.Backend.ReadQuorum = xl.metaReadQuorum
```
This is actually meaningless now, we should fix it and remove these variables. Printing them doesn't make sense unless we do it correctly. This also doesn't signify "default" - default is StandardClass which is also configurable.

fwessels replied:
Yes, agreed, let's remove these variables.

fwessels commented on cmd/xl-v1-object.go:
```diff
@@ -60,7 +60,15 @@ func (xl xlObjects) prepareFile(bucket, object string, size int64, onlineDisks [
+
+	// get Quorum for this object
+	readQuorum, writeQuorum, err := objectQuorumFromMeta(xl, metaArr, errs)
+
```
Spurious new line

fwessels commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,213 @@
+			parity = globalStandardStorageClass.Parity
+		} else {
+			// else fall back to default value
+			parity = totalDisks / 2
```
You could initialize parity around line 146 to `totalDisks / 2`, and then remove default cases in lines 162 and 165

fwessels commented on cmd/storage-class_test.go:
```diff
@@ -0,0 +1,280 @@
+		{2, 6, 5, nil},
+		{3, 1, 0, errors.New("Standard storage class parity disks should be greater than or equal to 2")},
+		{4, 4, 6, errors.New("Standard storage class parity disks should be greater than 6")},
+		{5, 9, 0, errors.New("Standard storage class parity disks should be less than or equal to 8")},
```
Maybe add a test case where ssParity and rrsParity are equal?

fwessels commented on cmd/storage-class_test.go:
```diff
@@ -0,0 +1,280 @@
+	}{
+		{1, *xl, parts, errs, 8, 9, nil},
+		{2, *xl, parts1, errs1, 14, 15, nil},
+		{3, *xl, parts2, errs2, 8, 9, nil},
```
For this test, use different value for standardStorageClass? (now you can not differentiate if it is actually different from standard case).
Ie if set to 6, then outcome should be 10 and 11 ?

fwessels commented on cmd/test-utils_test.go:
```diff
@@ -1999,6 +2010,31 @@ func ExecObjectLayerTest(t TestErrHandler, objTest objTestType) {
 	defer removeRoots(append(fsDirs, fsDir))
 }
 
+// ExecObjectLayerTest - executes object layer tests.
```
function name is incorrect.

fwessels commented on cmd/xl-v1-healing-common.go:
```diff
@@ -121,6 +121,27 @@ func listOnlineDisks(disks []StorageAPI, partsMetadata []xlMetaV1, errs []error)
+	}
+	// This is a slice with either partsMetaData that were updated the latest
+	// or empty partsMetaData
+	return latestXLMeta
```
Again (see earlier comment), return count of common xl.json's

fwessels commented on cmd/xl-v1-object.go:
```diff
@@ -154,7 +162,15 @@ func (xl xlObjects) GetObject(bucket, object string, startOffset int64, length i
+
+	// get Quorum for this object
+	readQuorum, _, err := objectQuorumFromMeta(xl, metaArr, errs)
+
```
Idem

fwessels commented on cmd/xl-v1-object.go:
```diff
@@ -659,6 +681,16 @@ func (xl xlObjects) deleteObject(bucket, object string) error {
+
+	// get Quorum for this object
+	_, writeQuorum, err := objectQuorumFromMeta(xl, metaArr, errs)
+
```
Idem

fwessels commented on cmd/xl-v1.go:
```diff
@@ -145,9 +142,9 @@ func newXLObjects(storageDisks []StorageAPI) (ObjectLayer, error) {
-	xl.writeQuorum = writeQuorum
+	// READ Quorum is always set to (N/2) and WRITE quorum is (N/2 +1). N is the number of disks for metadata files.
+	xl.metaReadQuorum = readQuorum
+	xl.metaWriteQuorum = writeQuorum
```
Move lines up into construction of xlObjects (lines 111 - 133)?

fwessels replied:
@harshavardhana Can we get rid of defining metaWriteQuorum separately? 
And instead define a contant like "WriteOverReadQuorum" to derive it from a read quorum? Then we can use the same constant in `storage-class.go`, line 212

By default it would be 1, but in certain cases increasing it can give additional Put security.

harshavardhana replied:
Yeah I was thinking we don't need metaWriteQuorum since there is no reason for these variables.

harshavardhana commented on cmd/xl-v1-healing-common.go:
```diff
@@ -184,7 +207,11 @@ func xlHealStat(xl xlObjects, partsMetadata []xlMetaV1, errs []error) HealObject
 	modTime, count := commonTime(listObjectModtimes(partsMetadata, errs))
-	if count < xl.readQuorum {
+
+	// get Quorum for this object
```
get read quorum for this object.

harshavardhana commented on cmd/xl-v1-healing.go:
```diff
@@ -81,13 +81,15 @@ func (xl xlObjects) HealBucket(bucket string) error {
 		return err
 	}
 
+	writeQuorum := len(xl.storageDisks)/2 + 1
```
Add a comment..

harshavardhana commented on cmd/xl-v1-healing.go:
```diff
@@ -535,6 +531,15 @@ func healObject(storageDisks []StorageAPI, bucket, object string, quorum int) (i
+	// Read metadata files from all the disks
+	partsMetadata, errs := readAllXLMetadata(xl.storageDisks, bucket, object)
+
+	// get Quorum for this object
```
get read quorum...

harshavardhana commented on cmd/xl-v1-healing_test.go:
```diff
@@ -333,8 +333,12 @@ func TestQuickHeal(t *testing.T) {
+
 	// Heal the missing buckets.
-	if err = quickHeal(xl.storageDisks, xl.writeQuorum, xl.readQuorum); err != nil {
+	if err = quickHeal(*xl, writeQuorum, readQuorum); err != nil {
```
You could simply take `xl` directly and change the `quickHeal(xlObj *xlObjects ...`

harshavardhana commented on cmd/xl-v1-healing-common.go:
```diff
@@ -184,7 +207,11 @@ func xlHealStat(xl xlObjects, partsMetadata []xlMetaV1, errs []error) HealObject
-	if count < xl.readQuorum {
+
+	// get Quorum for this object
+	readQuorum, _, _ := objectQuorumFromMeta(xl, partsMetadata, errs)
```
Is it okay to ignore err here?

harshavardhana replied:
Not sure if you saw this @nitisht ?

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -160,7 +147,7 @@ func newXLObjects(storageDisks []StorageAPI) (ObjectLayer, error) {
 
 	// Perform a quick heal on the buckets and bucket metadata for any discrepancies.
-	if err = quickHeal(xl.storageDisks, xl.writeQuorum, xl.readQuorum); err != nil {
+	if err = quickHeal(*xl, writeQuorum, readQuorum); err != nil {
```
Just take the `xl` directly no pointer deference

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -264,7 +251,5 @@ func getStorageInfo(disks []StorageAPI) StorageInfo {
 // StorageInfo - returns underlying storage statistics.
 func (xl xlObjects) StorageInfo() StorageInfo {
 	storageInfo := getStorageInfo(xl.storageDisks)
-	storageInfo.Backend.ReadQuorum = xl.readQuorum
```
Perhaps we can get rid of ReadQuorum checks from server-startup-msg.go and also from the StorageInfo.Backend struct and also please check this with @abperiasamy on the final banner requirement.

nitisht replied:
good catch. updated the banner as discussed with @abperiasamy

krisis commented on cmd/storage-class_test.go:
```diff
@@ -0,0 +1,355 @@
+		{7, *xl, parts7, errs7, 14, 15, nil},
+	}
+	for _, tt := range tests {
+		data, parity, err := objectQuorumFromMeta(tt.xl, tt.parts, tt.errs)
```
objectQuorumFromMeta returns read and write quorum values when err is nil. It is confusing to call the return value as data and parity. we could call them `actualReadQuorum` and `actualWriteQuorum`.

harshavardhana commented on cmd/xl-v1-healing.go:
```diff
@@ -81,6 +81,7 @@ func (xl xlObjects) HealBucket(bucket string) error {
 		return err
 	}
 
+	// get write quorum for this object.
```
No a write quorum for an object, this function is HealBucket.

harshavardhana commented on cmd/server-startup-msg.go:
```diff
@@ -173,14 +177,37 @@ func getStorageInfoMsg(storageInfo StorageInfo) string {
+	return msg
+}
+
+func getRRSStorageClassInfoMsg(storageInfo StorageInfo) string {
```
Can you add tests for this?

nitisht replied:
sure will do

harshavardhana commented on cmd/object-api-datatypes.go:
```diff
@@ -43,10 +43,10 @@ type StorageInfo struct {
-		WriteQuorum  int // Minimum disks required for successful write operations.
+		OnlineDisks    int // Online disks during server startup.
+		OfflineDisks   int // Offline disks during server startup.
+		standardSCData int // Data disks for currently set Standard storage class.
```
/currently set/currently configured/

harshavardhana commented on cmd/xl-v1-healing-common.go:
```diff
@@ -184,7 +207,11 @@ func xlHealStat(xl xlObjects, partsMetadata []xlMetaV1, errs []error) HealObject
-	if count < xl.readQuorum {
+
+	// get read quorum for this object
+	readQuorum, _, _ := objectQuorumFromMeta(xl, partsMetadata, errs)
```
Not sure why are we ignoring error here..

nitisht replied:
I missed this comment initially, but since `objectQuorumFromMeta` does the same test before throwing the error as done below this statement, IMO we can ignore the error here.

harshavardhana replied:
We can clarify that here.

nitisht replied:
sure will do

nitisht replied:
actually you were right, we shouldn't ignore err here. Updated the PR

harshavardhana reviewed (dismissed):
LGTM

nitisht committed:
Add x-amz-storage-class support

This adds configurable data and parity options on a per object
basis. To use variable parity

- Users can set environment variables to cofigure variable
parity

- Then add header x-amz-storage-class to putobject requests
with relevant storage class values

Fixes #4997

fwessels commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,187 @@
+	// latestXLMeta is updated most recently.
+	// We implicitly assume that all the xlMeta(s) have same dataBlocks and parityBlocks.
+	// We now check that at least dataBlocks number of xlMeta is available. This means count
+	// should be greater than or equal to dataBlocks field of validXLMeta. If not we throw read quorum error.
```
validXLMeta --> lastXLMeta

fwessels commented on cmd/storage-class.go:
```diff
@@ -0,0 +1,187 @@
+	}
+
+	// Since all the valid erasure code meta updated at the same time are equivalent, pass dataBlocks from one
+	// file to get the quorum
```
from one file --> from latestXLMeta

fwessels commented on docs/erasure/storage-class/README.md:
```diff
@@ -0,0 +1,78 @@
+# Minio Storage Class Quickstart Guide [![Slack](https://slack.minio.io/slack?type=svg)](https://slack.minio.io)
+
+Minio server supports storage class in erasure coded mode. This allows configurable data and parity disks per object.
```
erasure coded -> erasure coding

fwessels commented on docs/erasure/storage-class/README.md:
```diff
@@ -0,0 +1,78 @@
+- Less than N/2, if SS parity is not set.
+- Less than SS Parity, if it is set.
+
+As parity below 2 is not recommended, RR storage class is not supported for 4 disks erasure code setup.
```
erasure code --> erasure coding

fwessels reviewed (dismissed):
LGTM -- nice work and great functionality/improvement.

(I will do some further testing offline but don't expect any issues)

harshavardhana reviewed (approved):
LGTM
